apps/gk_field_multib.c

```diff
@@ -0,0 +1,197 @@
+{
+
+  struct gk_field_multib *mbf = gkyl_malloc(sizeof(struct gk_field_multib));
+
+  mbf->info = gk_multib->field;
+
+  mbf->gkfield_id = mbf->info.gkfield_id ? mbf->info.gkfield_id : GKYL_GK_FIELD_ES;
+
+  int num_blocks = 0;
+  int crossz_blocks[GKYL_MAX_BLOCKS] = {-1};
+  set_crossz_idxs(mba, app, bidx, crossz_blocks, &num_blocks);
+
+  // We need first to create the cross z ranges.
+  // For the two-block case:
+  int lower[2];
+  int upper[2];
+  lower[0] = mba->decomp_intrab[0]->parent_range.lower[0];
+  upper[0] = mba->decomp_intrab[0]->parent_range.upper[0];
+  lower[1] = mba->decomp_intrab[0]->parent_range.lower[1];
+  // Add up all ranges in z direction
+  upper[1] = 0;
+  for ( int i = 0; i < num_blocks; i++) {
+    upper[1] += gkyl_range_shape(&mba->decomp_intrab[crossz_blocks[i]]->parent_range, 1);
+  }
+  struct gkyl_range crossz;
+  gkyl_range_init(&crossz, mba->cdim, lower, upper);
+  int nghost[2] = {1,1};
+  gkyl_create_ranges(&crossz, nghost, &mbf->crossz_ext, &mbf->crossz);
+
+  // Create the decomp and communicator from the mb app communicator
+  // Stack all the individual ranges together
+  int num_ranges = 0;
+  for ( int i = 0; i < num_blocks; i++) {
+    num_ranges += mba->decomp_intrab[crossz_blocks[i]]->ndecomp;
+  }
+  struct gkyl_range ranges[num_ranges];
+  int range_count = 0;
+  int num_zcells_before = 0;
+  for (int ib = 0; ib < num_blocks; ib++) {
+    struct gkyl_rect_decomp *decomp_i = mba->decomp_intrab[ib];
+    for (int id = 0; id < decomp_i->ndecomp; id++) {
+      lower[0] = decomp_i->ranges[id].lower[0];
+      upper[0] = decomp_i->ranges[id].upper[0];
+      lower[1] = decomp_i->ranges[id].lower[1] + num_zcells_before;
+      upper[1] = decomp_i->ranges[id].upper[1] + num_zcells_before;
+      gkyl_range_init(&ranges[range_count], mba->cdim, lower, upper);
+      range_count+=1;
+    }
+    num_zcells_before += gkyl_range_shape(&decomp_i->parent_range, 1);
+  }
+
+  mbf->zdecomp = gkyl_rect_decomp_new_from_ranges(mba->cdim, ranges, num_ranges, &mbf->crossz);
+
+  mbf->zcomm = gkyl_comm_split_comm(mba->comm_multib, 0, mbf->zdecomp);
+
+  // Now get the sub range intersects
+  // Create global subrange we'll copy the field solver solution from (into local).
+  int rank;
+  gkyl_comm_get_rank(mbf->zcomm, &rank);
+  int intersect = gkyl_sub_range_intersect(&mbf->crossz_sub_range, &mbf->crossz, &app->global);
+
+  // allocate arrays for charge density
+  mbf->rho_c_global_dg = mkarr(mba->use_gpu, app->confBasis.num_basis, mbf->crossz_ext.volume);
+  mbf->rho_c_global_smooth = mkarr(mba->use_gpu, app->confBasis.num_basis, mbf->crossz_ext.volume);
+
+  // allocate arrays for electrostatic potential
+  mbf->phi = mkarr(mba->use_gpu, app->confBasis.num_basis, mbf->crossz_ext.volume);
+
+  mbf->phi_host = mbf->phi;
+  if (mba->use_gpu) {
+    mbf->phi_host = mkarr(false, app->confBasis.num_basis, mbf->crossz_ext.volume);
+  }
```

Comment:
1. please use phi_ho
2. I think a better way to do allocation of gkyl_arrays when you are accommodating CPU and GPU copies is
```
  mbf->phi = mkarr(mba->use_gpu, app->confBasis.num_basis, mbf->crossz_ext.volume);
  mbf->phi_ho = use_gpu? mkarr(false, mbf->phi_ho->ncomp, mbf->phi_ho->size)
                                            : gkyl_array_acquire(mbf->phi_ho);
```
and then in the release method use (independently of mbf->use_gpu)
```
gkyl_array_release(mbf->phi);
gkyl_array_release(mbf->phi_ho);
```
Please do it like this, here and in the future